test/jit/test_tracer.py

```diff
@@ -2356,7 +2356,10 @@ def forward(self, input_map: Dict[str, List[Tensor]]) -> Tensor:
 
         input_map = {"1" : [torch.rand(2, 2), torch.rand(2, 2)], "3" : [torch.rand(2, 2), torch.rand(2, 2)]}
         model = testA()
-        traced_model = torch.jit.trace(model, input_map)
+        example_input = list()
+        example_input.append(input_map)
+        example_input = tuple(example_input)
+        traced_model = torch.jit.trace(model, example_inputs=example_input)
```

Comment:
Can we avoid breaking this type of use case? This same usage would break for other users with similar usage patterns (i.e. `trace(model, example_inputs={'forward': dict})`).

Instead, can we provide a flag that disables the `dict` -> `tuple(dict)` conversion? e.g. `torch.jit.trace(model, inputs, unpack_input_dict=True)`. And `unpack_input_dict` would default to `False`.

Reply:
Hi, @davidberard98 
Thanks for your review!
For this use case, I personally think it is not appropriate. My protest opinion comes from below two point:
1. The current PyTorch documentation for [torch.jit.trace()](https://pytorch.org/docs/stable/generated/torch.jit.trace.html) specifies that the **example_inputs** parameter should come with the type of _tuple_ and _torch.Tensor_. I'm not sure these two type for **example_inputs** argument is a must or just a suggestion. If it is a must, this usage case shouldn't been included at first. At least, this undocumented behavior shouldn't been view as a standard.
2. Our solution is for the problem of how to pass variable args(positional) and kwargs in torch.jit.trace() for a given module. In this respective, a tuple is used for a packed of positional args while dict is used for pass keyword arguments, and dict used by this example should be viewed as an element of tuple since tracer support tuple with nested collection types. So, our dict is not in the same dimension with the this one.

For your suggestion with an additional argument to address backward compatibility:
1. I don't know if it is important to maintain this undocumented behavior for PyTorch with such a flag.
2. With this design of API, we need educate the user that the dict would substantially comes with two different meanings and the semantic of **example_inputs** parameter should depend on another one, that is, **unpack_input_dict**. For my opinion, making the semantics of different parameters as orthogonal as possible will give users a cleaner API.
3.  Inside the [tracer](https://github.com/pytorch/pytorch/pull/81623/files#diff-09665607e34394eac04617a39929e7ca9b732af62eb39dad48b1e47b733b2885L341), we have to deal with the case or a single tensor and dict. If we give user a unified front end interface, maybe we don't have to deal with such corner case.

Anyway, above is just my opinion and I think the decision of how to design the interface is up to PyTorch team.

Reply:
Hi, @davidberard98 
Thanks for your review!
For this use case, I personally think it is not appropriate. My protest opinion comes from below two points:
1. The current PyTorch documentation for [torch.jit.trace()](https://pytorch.org/docs/stable/generated/torch.jit.trace.html) specifies that the **example_inputs** parameter should come with the type of _tuple_ or _torch.Tensor_. I'm not sure these two type for **example_inputs** argument is a must or just a suggestion. If it is a must, this usage case shouldn't been included at first. At least, this undocumented behavior shouldn't been view as a standard.
2. Our solution is for the problem of how to pass variable args(positional) and kwargs in torch.jit.trace() for a given module. In this respective, a tuple is used for a packed of positional args while dict is used for pass keyword arguments, and dict used by this example should be viewed as an element of tuple since tracer support tuple with nested collection types. So, our dict is not in the same dimension with the this one.

For your suggestion with an additional argument to address backward compatibility:
1. I don't know if it is important to maintain this undocumented behavior for PyTorch with such a flag.
2. With this design of API, we need educate the user that the dict would substantially comes with two different meanings and the semantic of **example_inputs** parameter should depend on another one, that is, **unpack_input_dict**. For my opinion, making the semantics of different parameters as orthogonal as possible will give users a cleaner API.
3.  Inside the [tracer](https://github.com/pytorch/pytorch/pull/81623/files#diff-09665607e34394eac04617a39929e7ca9b732af62eb39dad48b1e47b733b2885L341), we have to deal with the case or a single tensor and dict. If we give user a unified front end interface, maybe we don't have to deal with such corner case.

Anyway, above is just my opinion and I think the decision of how to design the interface is up to PyTorch team.

Reply:
@tangleintel I definitely see your point; however I discussed with some other pytorch maintainers who agree with me - i.e. even though this isn't documented, torch.jit.trace is widely used and we'd prefer to avoid breaking use cases that depend on the current behavior.

We can probably deprecate the old behavior with dict inputs, e.g. see [the 180 days + 2 releases policy](https://github.com/pytorch/pytorch/wiki/PyTorch's-Python-Frontend-Backward-and-Forward-Compatibility-Policy). However I'd probably need your help pushing an update after 180 days, since I'd probably forget :)

Reply:
@davidberard98 
OK, then let's follow your suggestion to maintain the backward compatibility and follow the Pytorch's BC-breaking rules to mark the undocumented behavior as deprecated. Is it OK for you?
BTW, according to the BC-breaking rules, if we want to deprecate an old behavior, we need specify which pytorch version the deprecation first takes effect in a warning msg. So, which version you think this feature can be landed in(and the deprecated in effect version is at least plus 2 right?) ?
 Yes, we will consistently track this feature and give the update when time is right.